feat: add configurable timeout for unsent messages (#944)

jonas-jonas opened the pull request:
<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.

This text will be included in the changelog. If applicable, include links to documentation or pieces of code.
If your change includes breaking changes please add a codeblock documenting the breaking change:

```
BREAKING CHANGES: This patch changes the behavior of configuration item `foo` to do bar. To keep the existing
behavior please do baz.
```
-->
This PR adds the option to configure a TTL (Time-To-Live) for messages that can not be delivered. After the TTL has passed since the message's creation and it was not delivered, the message is `Abandoned` and the courier will not try to deliver the message again. This is done by setting its `status` to `MessageStatusAbandoned` in the database. The courier will only try to deliver messages with the status `MessageStatusQueued`. 

## Related issue(s)

This PR resolves https://github.com/ory/kratos/issues/944

<!--
If this pull request

1. is a fix for a known bug, link the issue where the bug was reported in the format of `#1234`;
2. is a fix for a previously unknown bug, explain the bug and how to reproduce it in this pull request;
2. implements a new feature, link the issue containing the design document in the format of `#1234`;
3. improves the documentation, no issue reference is required.

Pull requests introducing new features, which do not have a design document linked are more likely to be rejected and take on average 2-8 weeks longer to
get merged.

You can discuss changes with maintainers either in the Github Discusssions in this repository or
join the [Ory Chat](https://www.ory.sh/chat).
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR.

Please be aware that pull requests must have all boxes ticked in order to be merged.

If you're unsure about any of them, don't hesitate to ask. We're here to help!
-->

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](docs/docs).

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ory/kratos?pullRequest=2257) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ory/kratos?pullRequest=2257) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ory/kratos?pullRequest=2257) it.</sub>

jonas-jonas commented:
Hi @Benehiko

thanks for the detailed answer. I have added a few tests to the relevant classes. I also added the key to the `config.schema.json` file. 

This should be everything and the PR is ready-for-review.

Benehiko commented on embedx/config.schema.json:
```diff
@@ -1394,6 +1394,17 @@
             "/conf/courier-templates"
           ]
         },
+        "message_ttl": {
+          "description": "Defines a Time-To-Live for courier messages that could not be delivered. After the defined TTL has expired for a message that message is abandoned.",
+          "type": "string",
+          "pattern": "^([0-9]+(ns|us|ms|s|m|h))+$",
+          "default": "1h",
+          "examples": [
+            "1h",
+            "1m",
+            "1s"
+          ]
+        },
         "smtp": {
```
Nice! :)

Benehiko reviewed (changes requested):
This looks very thorough! Very nicely done :)

I have a few comments to improve this PR:
- We need a check in the persistence layer to not query abandoned projects. Somewhere [here](https://github.com/ory/kratos/blob/master/persistence/sql/persister_courier.go#L27-L39). This also means a corresponding test to make sure the `courier.NextMessages` function does not return abandoned messages.
- We need configuration tests [here](https://github.com/ory/kratos/blob/master/driver/config/config_test.go) and [here](https://github.com/ory/kratos/tree/master/test/schema). 

Other than that I think this is quite close to being merged

jonas-jonas commented:
> We need a check in the persistence layer to not query abandoned projects. Somew[here](https://github.com/ory/kratos/blob/master/persistence/sql/persister_courier.go#L27-L39) here. This also means a corresponding test to make sure the courier.NextMessages function does not return abandoned messages.

That method already only queries messages that have the `status=MessageStatusQueued`. I think that should exclude messages that have the `status= MessageStatusAbandoned` (or any other `MessageStatus` for that matter). This is also tested [here](https://github.com/ory/kratos/blob/b1f270c12e7193f3d2abb65947ed92c1906ee69c/courier/test/persistence.go#L93-L95) by manually setting the `status` of a message to Abandoned and then checking the result to be empty. Am I missing something here?


>We need configuration tests [here](https://github.com/ory/kratos/blob/master/driver/config/config_test.go) and [here](https://github.com/ory/kratos/tree/master/test/schema).

I will add the tests later today.

Benehiko commented:
> > We need a check in the persistence layer to not query abandoned projects. Somew[here](https://github.com/ory/kratos/blob/master/persistence/sql/persister_courier.go#L27-L39) here. This also means a corresponding test to make sure the courier.NextMessages function does not return abandoned messages.
> 
> That method already only queries messages that have the `status=MessageStatusQueued`. I think that should exclude messages that have the `status= MessageStatusAbandoned` (or any other `MessageStatus` for that matter). This is also tested [here](https://github.com/ory/kratos/blob/b1f270c12e7193f3d2abb65947ed92c1906ee69c/courier/test/persistence.go#L93-L95) by manually setting the `status` of a message to Abandoned and then checking the result to be empty. Am I missing something here?
> 
> > We need configuration tests [here](https://github.com/ory/kratos/blob/master/driver/config/config_test.go) and [here](https://github.com/ory/kratos/tree/master/test/schema).
> 
> I will add the tests later today.

Ah yes, sorry about that. You are correct :)

jonas-jonas commented:
I added the config tests. This should be everything.

zepatrik commented on courier/test/persistence.go:
```diff
@@ -89,6 +89,10 @@ func TestPersister(ctx context.Context, newNetworkUnlessExisting NetworkWrapper,
 			_, err = p.NextMessages(ctx, 1)
 			require.ErrorIs(t, err, courier.ErrQueueEmpty)
+
+			require.NoError(t, p.SetMessageStatus(ctx, messages[0].ID, courier.MessageStatusAbandoned))
```
Ah, I see. The persister already only gets pending ones.

zepatrik reviewed (approved):
Nice, looks good to me :+1: 
Waiting for @Benehiko to confirm.

Benehiko reviewed (approved):
LGTM :)

Really nicely done!

vinckr commented:
Hello @jonas-jonas 
Congrats on merging your first PR in Ory 🎉 !
Your contribution will soon be helping secure millions of identities around the globe 🌏. 
As a small token of appreciation we send all our first time contributors a gift package to welcome them to the community. 
[Please drop me an email and I will forward you the form to claim your Ory swag!](mailto:vincent@ory.sh)
